### ISSUE TITLE
Pyenv virtualenv listing and activation

### ISSUE THREAD

**Issue by rakanalh** [jorgenschaefer/elpy#803]
Hello,

Is there a way to setup elpy to only use pyenv & pyenv virtualenv?
https://github.com/yyuu/pyenv-virtualenv

This means that detecting the current python version from pyenv local/global configs and using ~/.pyenv/versions/{Version}/envs 
To list available virtualenvs.

Is this currently possible in elply?

Many Thanks

**Comment by jorgenschaefer** [comment]
Hello, and thanks for the question/suggestion – that's not possible right now, but it would be a great addition. :-)

**Comment by gregnewman** [comment]
My setup is pretty much like that.  Look in I talked packages and python
sections. I'm on my phone so can point to line numbers. Sorry.

https://github.com/gregnewman/emacs-config/blob/master/greg.org

I'm using pyenv-mode with elpy and it automatically reads the python
version.

**Comment by gregnewman** [comment]
Look in installed packages. Sorry for the typos.

**Comment by jorgenschaefer** [comment]
Nice!

Incidentally, when you set `WORKON_HOME` (e.g. using `(setenv "WORKON_HOME" "...")`), `M-x pyvenv-workon` will pick up the virtualenvs in that directory.

**Comment by rakanalh** [comment]
Thanks alot. I'll look into your suggestions and report what i've been able to setup accordingly.

**Comment by jorgenschaefer** [comment]
Great. If you come up with a nice generic solution, I'd be quite interested in getting it into Elpy :-)

**Comment by rakanalh** [comment]
Thanks, but i am no elisp expert so i am going to need some hand holding to get there :).

Here's what i've come up with based on @gregnewman's page and your suggestion for WORKON_HOME:
```
(defun ssbb-pyenv-hook ()
"Automatically activates pyenv version if .python-version file exists."
(f-traverse-upwards
 (lambda (path)
    (let ((pyenv-version-path (f-expand ".python-version" path)))
     (if (f-exists? pyenv-version-path)
         (progn
           (setq python-current-version (s-trim (f-read-text pyenv-version-path 'utf-8)))
           (pyenv-mode-set python-current-version)
           (setenv "WORKON_HOME" (concat "~/.pyenv/versions/" python-current-version "/envs"))
           (message (concat "Setting virtualenv path to ~/.pyenv/versions/" python-current-version "/envs"))))))))

(add-hook 'find-file-hook 'ssbb-pyenv-hook)
```

The problems with this approach are:

1. Switching between files in multiple projects that may use different python versions is not accounted for. This will use the last project's python version.
2. If we activate a project with a specific version then activate one of it's virtualenvs, switching to a new project with some other python version will not deactivate that virtualenv and switch to the new version.

I am not sure how heavy it is to do this on every buffer switch (between different projects) or otherwise, this can support one project at a time basis.

What do you guys think?

**Comment by jorgenschaefer** [comment]
Automatic switching of virtualenvs is kind of slow. pyvenv has a tracking mode to do that, but it can slow down interaction quite a bit. So maybe not for everyone. But you can probably hook into that.

**Comment by rakanalh** [comment]
Yup that's what i thought. For me at least it's not necessary to have something like that, but rather wanted to point out the down sides of this approach. It works well so far.

Thanks for helping :) :+1:

**Comment by jorgenschaefer** [comment]
You are very welcome! :-)

**Comment by michaelsteigman** [comment]
Coincidentally, I am in the same boat (and I know Greg, who shared his tip with me as well).

Wanted to add that the code to pick up .python-version above is now in an emacs package called pyenv-mode-auto (Melpa bleeding edge). I don't believe it looks in WORKON_HOME, however. 

Also, I've noticed that, for me at least, the auto-activation doesn't work properly with either the script above or the module. Or, I guess I should say that emacs reports that the right pyenv is active but I get "Flymake: Configuration error has occurred while running (flake8 /Users/ms/bar_flymake.py). Flymake will be switched OFF" error. If I open up elpy-config after hitting "OK", everything looks alright (to me at least):

Virtualenv........: lip (/Users/ms/.pyenv/versions/3.5.1/envs/lip)
RPC Python........: 3.5.1 (/Users/ms/.pyenv/shims/python)
Interactive Python: python (/Users/ms/.pyenv/shims/python)
Emacs.............: 24.5.1
Elpy..............: 1.11.0
Jedi..............: 0.9.0
Rope..............: Not found (0.10.3 available)
Importmagic.......: 0.1.3
Autopep8..........: 1.2.1
Yapf..............: 0.6.2
Syntax checker....: flake8 (/Users/ms/.pyenv/shims/flake8)

However, if I do a 'pyenv-set-version foo' before opening up a python file, everything is hunky dory. Any suggestions for debugging this issue? 

Thanks (and thanks for Elpy!)

**Comment by jorgenschaefer** [comment]
The problem sounds like as if flymake is activated before pyenv has a chance of setting up the environment there. I do not know how pyenv hooks itself up, though. :-(

**Comment by ejmg** [comment]
Unless I misunderstand the request, @jorgenschaefer's suggestion of simply setting `WORKON_HOME` within your init.el works for me in terms of sourcing the virtualenvs used by `pyenv`

Here's exactly what my setting looks like in my emacs config (I use `'use-package`, but these would work all the same if put in a normal init.el file):

```
(use-package elpy
   :config
   (package-initialize)
   (elpy-enable)
   (setenv "WORKON_HOME" "~/.pyenv/versions/")
   (setq elpy-rpc-backend "jedi")
   (setq python-shell-interpreter "~/.pyenv/shims/python3")
 )
```

**Comment by ejmg** [comment]
Building on my comment, assuming the appropriate `virtualenv` can be activated, doesn't that handle the problem of what version of Python is being used by Elpy (minus the interpreter for the inferior shell)?

**Comment by jorgenschaefer** [comment]
> minus the interpreter for the inferior shell

Even that, if the variable does not contain a full path and the interactive interpreter is installed in the virtualenv.
